railgun/internal/cmd/rootcmd/servers.go

```diff
@@ -45,23 +45,32 @@ func StartAdmissionServer(ctx context.Context, c *manager.Config) error {
 	return nil
 }
 
-func StartProfilingServer(ctx context.Context, c *manager.Config) error {
+func StartDiagnosticsServer(ctx context.Context, port int, c *manager.Config) (diagnostics.Server, error) {
 	deprecatedLogger, err := util.MakeLogger(c.LogLevel, c.LogFormat)
 	if err != nil {
-		return err
+		return diagnostics.Server{}, err
 	}
 	logger := logrusr.NewLogger(deprecatedLogger)
 
-	if !c.EnableProfiling {
-		logger.Info("profiling server disabled")
-		return nil
+	if !c.EnableProfiling && !c.EnableConfigDumps {
+		logger.Info("diagnostics server disabled")
+		return diagnostics.Server{}, nil
 	}
 
-	s := diagnostics.Server{Logger: logger}
+	s := diagnostics.Server{
+		Logger:           logger,
+		ProfilingEnabled: c.EnableProfiling,
+	}
+	if c.EnableConfigDumps {
+		s.ConfigDumps = util.ConfigDumpDiagnostic{
+			DumpsIncludeSensitive: c.DumpSensitiveConfig,
+			Configs:               make(chan util.ConfigDump, 3),
```

Comment:
why a buffered channel with 3 items ?

Reply:
See [Shane's comment from earlier](https://github.com/Kong/kubernetes-ingress-controller/pull/1589#discussion_r676904718). The buffer allows this to be non-blocking with a small amount of leeway to continue shipping updated configuration if the diagnostic server's receiver goroutine is too busy to ingest configuration, or drop diagnostic updates if it's so busy that it hits the buffer limit.

Size choice was essentially arbitrary. I wanted something that didn't immediately start dropping updates if the other goroutine wasn't ready yet, but figured it didn't need to be very large, since that routine does nothing but listen for updates. If field evidence indicates we should change it for some reason we can revisit the size choice in the future.

Reply:
well, all buffered channel is for asynchronous.
The size should be decided by the write frequency and the number of your consumer goroutine.